cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskPlanner.java

```diff
@@ -96,39 +112,111 @@ public void addExecutionProposals(Collection<ExecutionProposal> proposals,
                                     Cluster cluster,
                                     ReplicaMovementStrategy replicaMovementStrategy) {
     LOG.trace("Cluster state before adding proposals: {}.", cluster);
-    maybeAddReplicaMovementTasks(proposals, cluster, replicaMovementStrategy);
+    maybeAddInterBrokerReplicaMovementTasks(proposals, cluster, replicaMovementStrategy);
+    maybeAddIntraBrokerReplicaMovementTasks(proposals);
     maybeAddLeaderChangeTasks(proposals, cluster);
+    sanityCheckExecutionTasks();
+    maybeDropReplicaSwapTasks();
   }
 
   /**
-   * For each proposal, create a replica action task if there is a need for moving replica(s) to reach expected final proposal state.
+   * Sanity check that if there is any intra-broker partition movement task generated, no inter-broker partition movement task
+   * is generated.
+   */
+  private void sanityCheckExecutionTasks() {
+    if (_remainingIntraBrokerReplicaMovements.size() > 0) {
+      for (ExecutionTask task : _remainingInterBrokerReplicaMovements) {
+        if (task.proposal().replicasToAdd().size() > 0) {
+          throw new IllegalStateException("Intra-broker partition movement should not mingle with inter-broker partition movement.");
+        }
+      }
+    }
+  }
+
+  /**
+   * If there is any intra-broker partition movement task generated, any replica swap operation is cancelled.
+   * The reason to do this is because disk rebalance operation over a cluster with dead broker or disk may generate some
+   * unnecessary replica swap operations. These swap operations do not improve the disk load balance but bring a potential
+   * risk of hang(if the swap operation involves offline replica).
+   */
+  private void maybeDropReplicaSwapTasks() {
```

Comment:
The use of this method is not clear to me. 

> `If there is any intra-broker partition movement task generated, any replica swap operation is cancelled.`
Who creates these tasks? is it inter or intra broker swap?

Reply:
It is inter-broker replica swap operation(equivalently change the replica order in assigned replica list of the partition).
The reason why I have this method is that I observe in a cluster with dead disk, if the replica to swap is on dead disk, Kafka cannot finish the operation.

for example, there are 2 broker in the cluster:
broker 1 with disk1(dead), disk2(alive)
broker 2 with disk1(alive), disk2(alive)

An operation on a partition with original replica list (replica1(broker1,disk1), replica2(broker2, disk1)) and want to change to  (replica2(broker2, disk1), replica1(broker1,disk1) ) will hang because replica1 is offline.